Improved API documentation for trait_types

corranwebster opened the pull request:
General overhaul of docstrings in the `traits.trait_types` module.

- Moved parameters from __init__ to class docstrings
- Added class attributes.
- Cleaned up exposition where it was needed.

mdickinson commented on traits/trait_types.py:
```diff
@@ -149,11 +156,12 @@ def default_text_editor(trait, type=None):
 def _validate_int(value):
-    """
-    Convert an integer-like Python object to an int, or raise TypeError.
+    """ Convert an integer-like Python object to an int, or raise TypeError.
```
This seems a bit subjective. I personally prefer the original style here, but I reluctantly agree that this style is more consistent with the rest of the file.

corranwebster replied:
Yes, I'm not entirely happy that there are a number of places where there wasn't a really good way to get a nice one-liner for the docstring.  I personally prefer the
```
""" Brief summary on one line.

More detailed information
"""
```
and
```
""" Brief summary for trivial thing.
"""
```
(or with the triple quotes on the same line if it will fit).

mdickinson commented on traits/trait_types.py:
```diff
@@ -225,19 +226,13 @@ def create_editor(self):
 class Int(BaseInt):
-    """ Defines a trait whose type must be an integer using a C-level fast
-        validator.
+    """ Defines trait that holds an int and uses a C-level fast validator.
```
The wording of "integer" rather than "int" the original docstring was deliberate, since `Int` accepts integer-like things (most importantly, NumPy integers) as well as plain `int`s.  Maybe we should spell this out more fully? Something like: "must be an int or integer-like object", with a fuller explanation of what exactly "integer-like" means elsewhere.

mdickinson commented on traits/trait_types.py:
```diff
@@ -268,19 +263,13 @@ def create_editor(self):
 class Float(BaseFloat):
-    """ Defines a trait whose value must be a Python float using a C-level fast
-        validator.
+    """ Defines a trait that holds a float and uses a C-level fast validator.
```
Again, `float` is a bit misleading here. "float, or object adaptable to float"?

mdickinson commented on traits/trait_types.py:
```diff
@@ -314,19 +303,13 @@ def create_editor(self):
 class Complex(BaseComplex):
-    """ Defines a trait whose value must be a Python complex using a C-level
-        fast validator.
+    """ Defines a trait that holds a complex and uses a C-level fast validator.
```
Urg. "a complex" is horrible. Replace with "a complex number"? Or "an instance of `complex`"?

mdickinson commented on traits/trait_types.py:
```diff
@@ -361,17 +344,15 @@ def create_editor(self):
 class Str(BaseStr):
-    """ Defines a trait whose value must be a Python string using a C-level
-        fast validator.
+    """ Defines a trait that holds a str and uses a C-level fast validator.
```
replace "a str" with "a str object"? Or keep it as "a string"? Or even "a text string"?

mdickinson commented on traits/trait_types.py:
```diff
@@ -427,8 +403,7 @@ def create_editor(self):
 class Bytes(BaseBytes):
-    """ Defines a trait whose value must be a Python bytes string using a
-        C-level fast validator.
+    """ Defines a trait that holds a bytes and uses a C-level fast validator.
```
Suggest "bytes object" or "a bytestring" in place of "a bytes".

mdickinson commented on traits/trait_types.py:
```diff
@@ -472,22 +447,16 @@ def create_editor(self):
 class BaseCInt(BaseInt):
     """ Defines a trait whose value must be a Python int and which supports
-        coercions of non-int values to int.
+    coercions of non-int values to int.
```
For these cases, we should really try to come up with a one-liner that works, followed by a fuller explanation in a separate paragraph if necessary.

mdickinson commented on traits/trait_types.py:
```diff
@@ -1386,13 +1313,43 @@ class PythonValue(Any):
 class BaseFile(BaseStr):
     """ Defines a trait whose value must be the name of a file.
+
+    For Python 3.6 this will accept os.pathlib Path objects, converting them
```
Suggest adding "and later", or using "Python >= 3.6". (Also applies to the `File` docstring.)

corranwebster commented:
I think this addresses all the above issues; I've tried to re-word all of the first lines of the docstrings to be something like: "A fast-validating/coercing/... trait type whose value ..."

mdickinson commented on traits/trait_types.py:
```diff
@@ -192,13 +198,11 @@ class Any(TraitType):
-    """ Defines a trait whose type must be an integer.
+    """ A trait type whose value must be an int.
+
+    Values which support the Python index protocol will validated and will be
```
typo: "will validate"?

mdickinson replied:
Resolving this one; it's duplicated with the suggested change below.

mdickinson commented on traits/trait_types.py:
```diff
@@ -192,13 +198,11 @@ class Any(TraitType):
-    """ Defines a trait whose type must be an integer.
+    """ A trait type whose value must be an int.
+
+    Values which support the Python index protocol will validated and will be
```
```suggestion
    Values which support the Python index protocol will validate and will be
```

mdickinson commented on traits/trait_types.py:
```diff
@@ -225,21 +229,22 @@ def create_editor(self):
-        validator.
+    """ A fast-validating trait type whose value must be an integer.
+
+    Values which support the Python index protocol will validated and will be
```
```suggestion
    Values which support the Python index protocol will validate and will be
```

mdickinson commented on traits/trait_types.py:
```diff
@@ -268,21 +273,22 @@ def create_editor(self):
-    """ Defines a trait whose value must be a Python complex.
+    """ A trait type whose value must be a complex number.
+
+    Integers and floating point numbers will be converted to the
```
```suggestion
    Integers and floating-point numbers will be converted to the
```

mdickinson commented on traits/trait_types.py:
```diff
@@ -314,21 +320,18 @@ def create_editor(self):
-        fast validator.
+    """ A fast-validating trait type whose value must be a complex number.
+
+    Integers and floating point numbers will be converted to the
```
```suggestion
    Integers and floating-point numbers will be converted to the
```

mdickinson commented on traits/trait_types.py:
```diff
@@ -129,12 +125,23 @@
-
 
 def default_text_editor(trait, type=None):
+    """ Returns a default text editor for a trait.
```
```suggestion
    """ Return a default text editor for a trait.
```

mdickinson commented on traits/trait_types.py:
```diff
@@ -361,17 +364,15 @@ def create_editor(self):
 class Str(BaseStr):
-    """ Defines a trait whose value must be a Python string using a C-level
-        fast validator.
+    """ A fast-validating trait type whose value must be a complex number..
```
```suggestion
    """ A fast-validating trait type whose value must be a complex number.
```

mdickinson commented on traits/trait_types.py:
```diff
@@ -1036,18 +944,14 @@ class Module(TraitType):
-        assigned. It has no default value. This trait is most often used in
-        conjunction with wildcard naming. See the *Traits User Manual* for
-        details on wildcards.
+    """ A trait type that behaves as to a standard Python attribute.
```
```suggestion
    """ A trait type that behaves as a standard Python attribute.
```

mdickinson commented on traits/trait_types.py:
```diff
@@ -3019,7 +3036,7 @@ def init_fast_validate(self):
 
 class Supports(Instance):
-    """ A traits whose value must support a specified protocol.
+    """ A trait type whose value is adatped to a specified protocol.
```
```suggestion
    """ A trait type whose value is adapted to a specified protocol.
```

mdickinson commented on traits/trait_types.py:
```diff
@@ -3186,20 +3206,29 @@ def resolve_default_value(self):
-#  'Event' trait:
-# -------------------------------------------------------------------------------
+class Event(TraitType):
+    """ A trait type that holds no value but can be set and listend to.
```
```suggestion
    """ A trait type that holds no value but can be set and listened to.
```

mdickinson commented:
Still LGTM. Will merge and then backport when CI completes.
